File: src/main/java/me/testaccount666/serversystem/commands/wrappers/AbstractCommandWrapper.java
package me.testaccount666.serversystem.commands.wrappers;

import me.testaccount666.serversystem.ServerSystem;
import me.testaccount666.serversystem.userdata.User;
import me.testaccount666.serversystem.userdata.UserManager;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;

import java.util.Optional;

public abstract class AbstractCommandWrapper {
    protected static Optional<User> resolveCommandUser(CommandSender commandSender) {
        if (!(commandSender instanceof Player player)) return Optional.of(UserManager.Companion.getConsoleUser());

        var cachedUserOptional = ServerSystem.getInstance().getRegistry().getService(UserManager.class).getUser(player);

        if (cachedUserOptional.isEmpty()) return Optional.empty();

        var cachedUser = cachedUserOptional.get();

        if (cachedUser.isOfflineUser()) return Optional.empty();

        return Optional.ofNullable((User) cachedUser.getOfflineUser());
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/commands/wrappers/AbstractCommandWrapper.kt
package me.testaccount666.serversystem.commands.wrappers

import me.testaccount666.serversystem.extensions.getService
import me.testaccount666.serversystem.userdata.User
import me.testaccount666.serversystem.userdata.UserManager
import me.testaccount666.serversystem.userdata.UserManager.Companion.consoleUser
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player

abstract class AbstractCommandWrapper {
    internal fun resolveCommandUser(commandSender: CommandSender): User? {
        if (commandSender !is Player) return consoleUser

        val cachedUser = getService<UserManager>().getUserOrNull(commandSender) ?: return null
        if (cachedUser.isOfflineUser) return null

        return cachedUser.onlineUser
    }
}
